[achievements] Pinning Functionality UI and Tiny Bug Fix

jiayuew1 opened the pull request:
To address issue #1123,
Split the codebase into separate HTML, CSS, and JS files for better maintainability
In the src/static/js/achievements.js, I have added the following functionalities.
1. Adding animation to the pin icon when clicked (e.g., rotation and color change)
2. Providing immediate visual feedback in the UI when an achievement is pinned/unpinned
3. Visually updating the achievements list in the pinned section without page reload

I've also fixed the issue where the HTML was not displaying the correct username (Issue #1040). It appears that the User Profile team updated the database to store the name instead of the username. I've updated the HTML to display the username accordingly. Additionally, I changed the URL to use the user ID instead of the username, since the username can be nullable and is less reliable.

jiayuew1 committed:
pinning

jiayuew1 committed:
formatting

jiayuew1 committed:
Update achievements.css

jiayuew1 committed:
Update user_achievements.html

jiayuew1 committed:
Update urls.py

jiayuew1 committed:
Update views.py

jiayuew1 committed:
Update user_achievements.html

jiayuew1 committed:
Update views.py

jiayuew1 committed:
Update urls.py

jiayuew1 committed:
Update urls.py

mhenry111 commented:
I pulled the latest commit into my test branch and ran chigame. The new organization of the code looks good, but none of the new pinning functionality mentioned in your comment worked for me. Am I supposed to be able to pin/unpin achievements?

jiayuew1 commented:
> I pulled the latest commit into my test branch and ran chigame. The new organization of the code looks good, but none of the new pinning functionality mentioned in your comment worked for me. Am I supposed to be able to pin/unpin achievements?

I've added the pinning functionality to views.py in PR #1097. Ben has approved it, so I expect it will be merged soon. Due to time constraints, I went ahead and created the PR—sorry for any extra work this may cause on your end. Once you update the views and URLs, everything should work as expected.

mhenry111 commented:
> 
> I've added the pinning functionality to views.py in PR #1097. Ben has approved it, so I expect it will be merged soon. Due to time constraints, I went ahead and created the PR—sorry for any extra work this may cause on your end. Once you update the views and URLs, everything should work as expected.

Thanks for the clarification! The pinning/unpinning functionality looks great. I only have a couple recommendations, which can be addressed in future issues:

1. update code to not allow pinning for achievements that haven't been earned yet, because currently when they appear on the pinned achievements page, it says 'Unlocked on Not Unlocked' for the date. 
2. adjust achievements cards so that the longer achievement names aren't split in the middle of a word like this:

lexicograph
er

mhenry111 reviewed (approved):
nicely organized and works with the pinning functionality of #1097; looks good to me!

jiayuew1 committed:
Update achievements.js

jiayuew1 committed:
Update achievements.js

jiayuew1 commented:
> > I've added the pinning functionality to views.py in PR #1097. Ben has approved it, so I expect it will be merged soon. Due to time constraints, I went ahead and created the PR—sorry for any extra work this may cause on your end. Once you update the views and URLs, everything should work as expected.
> 
> Thanks for the clarification! The pinning/unpinning functionality looks great. I only have a couple recommendations, which can be addressed in future issues:
> 
> 1. update code to not allow pinning for achievements that haven't been earned yet, because currently when they appear on the pinned achievements page, it says 'Unlocked on Not Unlocked' for the date.
> 2. adjust achievements cards so that the longer achievement names aren't split in the middle of a word like this:
> 
> lexicograph er

I've fixed the display error where "Unlocked on not unlocked" was shown—it now correctly displays "Not yet unlocked." I don't think we need to add a constraint on the achievement status, as users may want to pin achievements they haven't earned yet as goals in the pinned achievements panel. As for the second point, I'll convert it into an issue and address it in the future.

mhenry111 commented:
> 
> I've fixed the display error where "Unlocked on not unlocked" was shown—it now correctly displays "Not yet unlocked." I don't think we need to add a constraint on the achievement status, as users may want to pin achievements they haven't earned yet as goals in the pinned achievements panel. 

Oh I hadn't thought about it that way. That sounds good to me!

jiayuew1 committed:
Merge branch 'dev' into achievement/dynamic_template

jiayuew1 committed:
Merge branch 'dev' into achievement/dynamic_template

jiayuew1 committed:
Merge branch 'dev' into achievement/dynamic_template

abargher committed:
Merge branch 'dev' into achievement/dynamic_template

abargher reviewed (approved):
Looks great! Nice to have things split up into separate files.

abargher commented:
Issue Score: Excellent

Comments:
Great work! This is very well documented.

<!-- {"score": "excellent", "minor": false} -->
